Update CSR controllers & kubelet to respect signerName field

munnerz opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Adds support for the `signerName` field added to the CertificateSigningRequest resource as part of the changes described in https://github.com/kubernetes/enhancements/pull/1400 to:

* CSR approval controller
* CSR signer controller
* Kubelet TLS bootstrapping/certificate manager
* client-go's `RequestCertificate` helper function.

**Special notes for your reviewer**:

The top commit of this branch is the one relevant to this PR 😄 

This PR depends on:
* https://github.com/kubernetes/kubernetes/pull/86933
* https://github.com/kubernetes/kubernetes/pull/86476

**Does this PR introduce a user-facing change?**:
```release-note
ACTION REQUIRED: consumers of the 'certificatesigningrequests/approval' API must now grant permission to 'approve' CSRs for the 'signerName' specified on the CSR. More information on the new signerName field can be found at https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md#signers
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/1400
```

/cc @enj @liggitt @deads2k

Commit:
Add signerName field to CSR resource spec

Signed-off-by: James Munnelly <james.munnelly@jetstack.io>

Commit:
Add Certificate signerName admission plugins

munnerz commented:
I set release note to NONE as the first PR already adds:

* Add 'signerName' field to the CertificateSigningRequest API to allow for external signers to be built

And the admission PR has:

* ACTION REQUIRED: consumers of the 'certificatesigningrequests/approval' API must now grant permission to 'approve' CSRs for the 'signerName' specified on the CSR. More information on the new signerName field can be found in the KEP: https://github.com/kubernetes/enhancements/pull/1400

which seems sufficient to me 😄 

Is there some way to collect release notes specific to breaking changes in `client-go`? This PR changes the function signature of `RequestCertificate` to add `signerName` @nikhita

liggitt commented on pkg/controller/certificates/approver/sarapprove.go:
```diff
@@ -146,45 +145,8 @@ func appendApprovalCondition(csr *capi.CertificateSigningRequest, message string
-		return false
-	}
-	return true
+	return capihelper.IsKubeletClientCSR(x509cr, csr.Spec.Usages)
```
if signerName is non-nil, also check if it matches the apiserver-client-kubelet signerName?

munnerz replied:
Ah, yes. I also need to adjust this so that if it _is_ nil, we in-memory 'default' the value of signerName.

liggitt commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -89,6 +89,19 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
 func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
+	// Special-case handling here in case we are talking to an older apiserver.
+	if csr.Spec.SignerName == nil {
+		return nil
```
For nil, signerNames, I expected the same defaulting as in the apiserver, per the KEP

```go
csr, err := parse
switch {
  case err:
    legacy
  case iskubeletclient(csr):
    kubelet client
  case isclient(csr):
    client
  default:
    legacy
}
```

liggitt commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -89,6 +89,19 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
+	if csr.Spec.SignerName == nil {
+		return nil
+	}
+	// Only handle CSRs with the specific known signerNames.
```
for this, I expected a switch on the resolved (maybe defaulted above) signerName:

```go
switch signerName {
case capi.KubeletServingSignerName:
  // also make sure IsKubeletServingCert passes on the CSR
case capi.KubeAPIServerClientKubeletSignerName:
  // also make sure IsAPIServerClientKubelet passes on the CSR
case capi.KubeAPIServerClientSignerName:
  // also make sure IsAPIServerClient passes on the CSR
case capi.LegacyUnknownSignerName:
  // any additional validation?
default:
  // ignore other signerNames
  return nil
}
```

munnerz replied:
> // also make sure IsAPIServerClient passes on the CSR

This does not currently exist as we don't have a definition/any restrictions on `kube-apiserver-client` (only `kube-apiserver-client-kubelet`).

The subject restrictions mentioned in the KEP are noted as being enforced during *admission* and not during approval/signing (related: @mikedanese's question here: https://github.com/kubernetes/enhancements/pull/1400#discussion_r374399364).

Should we:

1) enforce in all of admission, approval and signing? The additional check seems redundant unless we want to encourage restricting `system:masters` usage across approvers/signers.
2) enforce in approval and signing only?

munnerz replied:
Ah, I see there are also additional restrictions to apply defined in the KEP:

```
Trust distribution: signed certificates must be honored as client-certificates by the kube-apiserver. The CA bundle is not distributed by any other means.
Permitted subjects - no subject restrictions, but approvers and signers may choose not to approve or sign. Certain subjects like cluster-admin level users or groups vary between distributions and installations, but deserve additional scrutiny before approval and signing. An admission plugin is available to restrict system:masters, but it is often not the only cluster-admin subject in a cluster.
Permitted x509 extensions - Non-critical extensions should be dropped.
Permitted key usages - must include []string{"client auth"}. Must not include key usages beyond []string{"digital signature", "key encipherment", "client auth"}
Expiration/cert lifetime - minimum of CSR signer or request. Sanity of the time is the concern of the signer.
CA bit allowed/disallowed - not allowed.
```

liggitt replied:
Yeah, define a helper function that encapsulates the extensions/usages/ca requirements and call it here before signing a CSR that requests that signer name.

munnerz commented:
@liggitt I've updated this to address your comments 😄

munnerz commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -89,64 +91,72 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
+	case capi.KubeletServingSignerName:
+		return capihelper.IsKubeletServingCSR(req, usages)
+	case capi.KubeAPIServerClientKubeletSignerName:
+		return capihelper.IsKubeletServingCSR(req, usages)
```
@liggitt if the CSR does not exactly match, what should the signing controller do? It could either:

1) override the values
2) reject the request/mark it as invalid somehow (we don't have this option yet)
3) display a warning to the user
4) do nothing at all

Right now with this change we do (4), and simply `return nil`.

munnerz replied:
Adding (3) by simply firing an Event and returning nil could work, but has the potential to cause a lot of events as we have no way to recognise the terminal state of the CR.

(2) is the best solution long-term for external signers IMO, but probably needs consideration in the KEP due to the awkward design of the CSR `conditions` structure.

(1) is an option too - it'd require duplicating/shuffling around some of this checking logic to allow for usages to be different to what's requested by the user. It may be non-obvious though, and it *isn't* something that's been done in the past (old logic *requires* usages to be specified as-per policy). Given it's a net-addition and goes beyond what was described in the KEP changes however, I feel like it might be a bit brash to add it in here.

liggitt replied:
Do 4 for now. I'm looking into how to let signers indicate permanent failure in the CSR status

Commit:
certificates: update controllers to understand signerName field

Signed-off-by: James Munnelly <james.munnelly@jetstack.io>

munnerz commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -89,37 +91,90 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
+func requestValidForSignerName(req *x509.CertificateRequest, usages []capi.KeyUsage, signerName string) bool {
+	// Only handle CSRs with the specific known signerNames.
+	switch signerName {
+	case capi.KubeletServingSignerName:
```
Should we check here that the CommonName == the `spec.username`, similar to in the approval plugin?

liggitt replied:
No, the nature of the certificate and signerName is distinct from authorization.

munnerz commented:
/remove-sig apps
/remove-sig cli
/remove-sig api-machinery
/remove-sig testing
/remove-area apiserver
/remove-area kubectl

enj commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -89,37 +91,90 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
+	for _, u := range usages {
+		switch u {
+		case capi.UsageDigitalSignature:
+		case capi.UsageKeyEncipherment:
```
nit: collapse the two cases and add a comment saying these a optional.

enj commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -89,37 +91,90 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
+		//  communicate to the user why the request has been refused.
+		return nil
+	}
+	cert, err := s.sign(x509cr, csr.Spec.Usages)
```
I do not understand how this signing action makes sense with the actual signer name validation.  For example, in the case of `KubeAPIServerClientSignerName`, how do we make the API server honor the signed cert?  I would think we would need multiple `signer` controllers (one for each signer name) with (possibly) distinct private keys/CAs.

munnerz replied:
The signing controller is configured to sign certificates using the apiserver's CA. Does the apiserver support specifying multiple roots of trust for verifying clients? If so, a user could configure a 'node CA' vs a 'user CA' if they wanted to revoke a certain subset of clients.

I think it does make sense to support specifying a CA to use per `kubernetes.io/*` signerName and @deads2k mentioned interest in this feature a little while ago too, but I'm uncertain whether it needs to be a pre-requisite for adding the `signerName` feature - if we add additional options to control this in future this can be handled via defaulting and it's not something outlined in the KEP currently.

enj replied:
The `kubernetes.io/kubelet-serving` signer maps to `--kubelet-certificate-authority` (because this CA is a distinct flag, we can have a distinct private key for it).  We expect this signer to only be valid for serving certificates, but if we only use a single private key, then we are relying solely on cert usage checks.  We have no cryptographic isolation.

`kubernetes.io/kube-apiserver-client` and `kubernetes.io/kube-apiserver-client-kubelet` are a bit harder to separate since all we have is `--client-ca-file`.  @deads2k I might be misremembering but I seem to recall you doing something fancy with `--tls-sni-cert-key` and kubelets connecting over a particular network via IPs - but maybe that was for serving certs only.

In general I think we lack a way today to say "honor this CA for client certs but only if they match this particular shape (i.e. kubelet client cert vs. regular user client cert)."

`kubernetes.io/legacy-unknown` we can basically ignore.

liggitt replied:
> For example, in the case of KubeAPIServerClientSignerName, how do we make the API server honor the signed cert?

That's the responsibility of the cluster administrator giving this controller the CA to use to sign and the apiserver the CA to use to verify.

> I would think we would need multiple signer controllers (one for each signer name) with (possibly) distinct private keys/CAs.

For this release, I think letting this signer continue to sign all the `kubernetes.io/*` names it understands is reasonable. If needed, we could allow subdivision of that similar to how we allow enabling/disabling specific controller loops. For example:
* `--cluster-signing-signer-names=*,-kubernetes.io/legacy-unknown`
* `--cluster-signing-signer-names=kubernetes.io/kube-apiserver-client-kubelet,kubernetes.io/kube-apiserver-client`

enj replied:
> For this release, I think letting this signer continue to sign all the `kubernetes.io/*` names it understands with a single CA is reasonable.

Yeah I think it is probably the best we can do for now.

enj commented on pkg/controller/certificates/signer/signer_test.go:
```diff
@@ -99,3 +103,156 @@ func TestSigner(t *testing.T) {
+
+// noncryptographic for faster testing
+// DO NOT COPY THIS CODE
+var insecureRand = rand.New(rand.NewSource(0))
```
Was the real `crypto/rand` slow enough to matter?

munnerz replied:
Ironically this bit of code was copied from another area of the repo... Shall I get rid of this? I doubt it makes any/much difference.

enj replied:
I am fine with removing it but do not feel strongly.

munnerz commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -152,7 +144,7 @@ func requestValidForSignerName(req *x509.CertificateRequest, usages []capi.KeyUs
 		return capihelper.IsKubeletServingCSR(req, usages)
 	case capi.KubeAPIServerClientKubeletSignerName:
-		return capihelper.IsKubeletServingCSR(req, usages)
+		return capihelper.IsKubeletClientCSR(req, usages)
```
@liggitt should we also be checking if the username on the CSR matches the commonName of the request here? We currently check it in the approval plugin (but not in defaulting).

liggitt replied:
no, the signerName is based only on the csr/usages, not on the submitter

enj commented:
/priority important-soon

enj commented:
Needs release note for the new cluster roles.

enj commented on pkg/controller/certificates/signer/signer_test.go:
```diff
@@ -99,3 +103,156 @@ func TestSigner(t *testing.T) {
+			},
+		},
+		{
+			name: "should do nothing if signerName is nil",
```
This should panic right?

enj replied:
(if the CSR was approved it should deference the `nil` and `panic`)

munnerz replied:
Correct - but it's now no longer possible/valid to pass a CSR to `handle` that sets a `nil` signerName (or at least, it is unsupported/unhandled). So I'm not sure if this is really a valid test case to include anymore, unless we add a check into `handle` for a nil signerName and return early explicitly?

With the CSR being *not* approved, it is also expected that this test case does nothing, although it is then effectively a dupe of the case above it (`should do nothing if not approved`).

munnerz replied:
Updated these tests. We no longer have a test with an explicitly `nil` signerName (and do not expect/support a CSR with a nil signerName to be passed to handle). I've also expanded out the test coverage here too with a few additional cases (as well as fixing the failures)

enj replied:
Not testing for `nil` is fine.  Since the code would `panic`, the worse case is a DOS against the signer.

enj commented on pkg/controller/certificates/signer/signer_test.go:
```diff
@@ -99,3 +103,156 @@ func TestSigner(t *testing.T) {
 	}
 }
+
+func TestHandle(t *testing.T) {
```
I think this test is incomplete (you are missing usages).  Running it locally:

```
=== RUN   TestHandle
=== RUN   TestHandle/should_sign_if_signerName_is_kubernetes.io/kube-apiserver-client
    TestHandle/should_sign_if_signerName_is_kubernetes.io/kube-apiserver-client: signer_test.go:121: expected one Update action but got 0
=== RUN   TestHandle/should_sign_if_signerName_is_kubernetes.io/kube-apiserver-client-kubelet
    TestHandle/should_sign_if_signerName_is_kubernetes.io/kube-apiserver-client-kubelet: signer_test.go:136: expected one Update action but got 0
=== RUN   TestHandle/should_sign_if_signerName_is_kubernetes.io/legacy-unknown
=== RUN   TestHandle/should_sign_if_signerName_is_kubernetes.io/kubelet-serving
    TestHandle/should_sign_if_signerName_is_kubernetes.io/kubelet-serving: signer_test.go:166: expected one Update action but got 0
=== RUN   TestHandle/should_do_nothing_if_an_unrecognised_signerName_is_used
=== RUN   TestHandle/should_do_nothing_if_not_approved
=== RUN   TestHandle/should_do_nothing_if_signerName_is_nil
--- FAIL: TestHandle (0.02s)
    --- FAIL: TestHandle/should_sign_if_signerName_is_kubernetes.io/kube-apiserver-client (0.01s)
    --- FAIL: TestHandle/should_sign_if_signerName_is_kubernetes.io/kube-apiserver-client-kubelet (0.00s)
    --- PASS: TestHandle/should_sign_if_signerName_is_kubernetes.io/legacy-unknown (0.00s)
    --- FAIL: TestHandle/should_sign_if_signerName_is_kubernetes.io/kubelet-serving (0.00s)
    --- PASS: TestHandle/should_do_nothing_if_an_unrecognised_signerName_is_used (0.00s)
    --- PASS: TestHandle/should_do_nothing_if_not_approved (0.00s)
    --- PASS: TestHandle/should_do_nothing_if_signerName_is_nil (0.00s)
FAIL
```

enj reviewed (approved):
Just need to fix the unit test I believe.

munnerz commented:
Squashed changes up until your last review into a single commit, and added an additional commit containing the changes after your review: https://github.com/kubernetes/kubernetes/pull/88246/commits/9254ebebd9f3e8e14ba8ed6a54306aa1504f8e6e

Back over to you 😄

munnerz commented:
/retest

munnerz commented:
/retest

enj commented:
/lgtm
/retest
/hold

@munnerz can you add a complete set of action required and release notes for this PR?  I think we plan to merge just this one which will implicitly merge the other two.

enj commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/util/certificate/csr/csr.go:
```diff
@@ -46,16 +46,17 @@ import (
 // certificate (pem-encoded). If there is any errors, or the watch timeouts, it
 // will return an error.
-func RequestCertificate(client certificatesclient.CertificateSigningRequestInterface, csrData []byte, name string, usages []certificates.KeyUsage, privateKey interface{}) (req *certificates.CertificateSigningRequest, err error) {
+func RequestCertificate(client certificatesclient.CertificateSigningRequestInterface, csrData []byte, name string, signerName string, usages []certificates.KeyUsage, privateKey interface{}) (req *certificates.CertificateSigningRequest, err error) {
```
update ensureCompatible() to make sure that if `new.Spec.SignerName` and `orig.Spec.SignerName` are both non-nil, they match (and add unit tests for the nil, matching, and mismatching cases)

liggitt commented:
one comment on the ensureCompatible() method, then lgtm

looks like there is one other in-tree use that creates CertificateSigningRequests in kubeadm:
https://github.com/kubernetes/kubernetes/blob/7587ab3ef3d40ff8d2f04e420395815de2e77cd6/cmd/kubeadm/app/phases/certs/renewal/apirenewer.go#L84-L94
called from
https://github.com/kubernetes/kubernetes/blob/7587ab3ef3d40ff8d2f04e420395815de2e77cd6/cmd/kubeadm/app/phases/certs/renewal/manager.go#L253-L276
called from
https://github.com/kubernetes/kubernetes/blob/7587ab3ef3d40ff8d2f04e420395815de2e77cd6/cmd/kubeadm/app/cmd/alpha/certs.go#L239-L246

open a follow-up issue for kubeadm to update those CSRs to choose a signer name for each certificate they request (cc @neolit123)

liggitt commented:
also, this is the PR we'll merge (which will include the other two), so include all the relevant release notes in this one

neolit123 commented:
thanks for the heads up. i've opened:
https://github.com/kubernetes/kubeadm/issues/2047

i think i digested some of the details, but have one question regarding the timeline of the kubeadm change (1.18 vs next):
- the admission is now going to be enabled by default and a missing signerName in the CSR spec would mean fallback to the legacy signer (from defaulting). however, the legacy signer _does not have auto-approval by the KCM_. does that mean that kubeadm code, as is, will now timeout at the subsequent `WaitForCertificate` call [here](
https://github.com/kubernetes/kubernetes/blob/4e79344501bfc5d3bc48caa12c41b9246d90850d/cmd/kubeadm/app/phases/certs/renewal/apirenewer.go#L104) ?

if true this means we need to apply this change to kubeadm before CF next week and for 1.18.
wondering how this affects third party distributions that use CSR and auto-aproval.

disclaimer: i wouldn't exclude that i'm misunderstanding something.

liggitt commented:
CSR objects that don't have a signerName get defaulted based on the format of the request. Requests shaped like node client or serving certificates get those signerNames assigned, so auto approval flows for node client certificates continue working uninterrupted.

neolit123 commented:
> CSR objects that don't have a signerName get defaulted based on the format of the request. Requests shaped like node client or serving certificates get those signerNames assigned, so auto approval flows for node client certificates continue working uninterrupted.

ok, thanks for the clarification.

liggitt commented:
I'll take https://github.com/kubernetes/kubernetes/pull/88246#discussion_r385781127 as a follow up, would like to get this soaking ahead of freeze next week 

/lgtm
/approve

enj commented:
/hold cancel

tedyu commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,113 @@ limitations under the License.
+		l.Insert(string(s))
+	}
+	r := sets.NewString()
+	for _, s := range right {
```
It seems there is no need to create the second Set.
Checking sizes of the slices and whether each of the second slice is contained in the first Set should suffice.

liggitt replied:
that would return true for `[a, b, c]` vs `[a, a, b]`

tedyu replied:
The following would achieve same semantics with one map.
```
@@ -117,13 +116,23 @@ func IsKubeletClientCSR(req *x509.CertificateRequest, usages []certificatesv1bet
 // equalUnsorted compares two []string for equality of contents regardless of
 // the order of the elements
 func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
-       l := sets.NewString()
+       var l map[string]int
+       l = make(map[string]int)
        for _, s := range left {
-               l.Insert(string(s))
+               k := string(s)
+               cnt := l[k]
+               l[k] = cnt + 1
        }
-       r := sets.NewString()
        for _, s := range right {
-               r.Insert(string(s))
-       }
-       return l.Equal(r)
+               k := string(s)
+               cnt, ok := l[k]
+               if !ok {
+                       return false
+               }
+               cnt--
+               if cnt == 0 {
+                       delete(l, k)
+               }
+       }
+       return len(l) == 0
 }
```

BenTheElder replied:
if we really want to be cheap:
```go
func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
	if len(left) != len(right) {
		return false
	}
	l := make(map[string]int, len(left))
	for _, s := range left {
		k := string(s)
		l[k] = l[k] + 1
	}
	for _, s := range right {
		k := string(s)
		c := l[k]
		if c == 0 {
			return false
		}
		l[k] = (c - 1)
	}
	return true
}
```
but we do use the sets API all over, I'm not sure if this change is desirable.

tedyu replied:
```
	if len(left) != len(right) {
		return false
	}
```
We shouldn't use the above check since there may be duplicate in either side.
```
		c := l[k]
		if c == 0 {
```
The decrement should be done before checking c against 0.

I prefer the formation I presented above.
Set is not expressive in this use case. We don't have to use it.

BenTheElder replied:
> We shouldn't use the above check since there may be duplicate in either side.

If we're not checking for _identical_ counts in each side, why even count in the map ...?

> The decrement should be done before checking c against 0.

No, if we decrement to zero then we've had an exact match. If we find an entry with zero on the next check we did not. I did test this code 🙃

tedyu replied:
Thanks @BenTheElder for looking at my code.

Before we dig deeper into the code, can we get agreement on whether the following are considered equal:

[a, b] vs [a, a, b]

Along the current code, they're equal.
That was why I said we shouldn't check the lengths of two sides.

tedyu replied:
@BenTheElder was right in that the value of the map shouldn't be the number of times we see the same key from the left.
Here is revised code which keeps the current semantics and only uses one map:
```
func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
	var l map[string]int
	l = make(map[string]int)
	for _, s := range left {
		k := string(s)
		l[k] = 1
	}
	count := 0
	for _, s := range right {
		k := string(s)
		val, ok := l[k]
		if !ok {
			// not equal if we haven't seen this key from the left
			return false
		}
		if val == 1 {
			count++
			// we don't want to delete the key since there may be more occurrence of this key in right
			l[k] = -1
		}
	}
	// if we have seen all the keys from the right, they're equal
	return len(l) == count
}
```
I can refine the comment in the code if there is question on how it works or whether it is correct.

Thanks

tedyu replied:
We can keep this on the back burner and look out for frequent calls to similar equality checks.

tedyu commented on pkg/controller/certificates/signer/signer.go:
```diff
@@ -89,37 +91,82 @@ func newSigner(caFile, caKeyFile string, client clientset.Interface, certificate
+	}
+	if !requestValidForSignerName(x509cr, csr.Spec.Usages, *csr.Spec.SignerName) {
+		// TODO: mark the CertificateRequest as being in a terminal state and
+		//  communicate to the user why the request has been refused.
```
Is an error return suitable for conveying that the CertificateRequest is in terminal state ?

liggitt replied:
We need a place to record it on the API object for that to be effective. I'm working on a well-defined condition to communicate that.

BenTheElder commented:
I don't think this change is necessary, the current code is very readable
and AIUI this will not be heavily called.

On Sat, Feb 29, 2020, 13:14 Ted Yu <notifications@github.com> wrote:

> *@tedyu* commented on this pull request.
> ------------------------------
>
> In pkg/apis/certificates/v1beta1/defaults.go
> <https://github.com/kubernetes/kubernetes/pull/88246#discussion_r386056037>
> :
>
> > +	if !equalUnsorted(requiredUsages, usages) {
> +		return false
> +	}
> +
> +	return true
> +}
> +
> +// equalUnsorted compares two []string for equality of contents regardless of
> +// the order of the elements
> +func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
> +	l := sets.NewString()
> +	for _, s := range left {
> +		l.Insert(string(s))
> +	}
> +	r := sets.NewString()
> +	for _, s := range right {
>
> @BenTheElder <https://github.com/BenTheElder> was right in that the value
> of the map shouldn't be the number of times we see the same key from the
> left.
> Here is revised code which keeps the current semantics and only uses one
> map:
>
> func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
> 	var l map[string]int
> 	l = make(map[string]int)
> 	for _, s := range left {
> 		k := string(s)
> 		l[k] = 1
> 	}
> 	count := 0
> 	for _, s := range right {
> 		k := string(s)
> 		val, ok := l[k]
> 		if !ok {
> 			// not equal if we haven't seen this key from the left
> 			return false
> 		}
> 		if val == 1 {
> 			count++
> 			// we don't want to delete the key since there may be more occurrence of this key in right
> 			l[k] = -1
> 		}
> 	}
> 	// if we have seen all the keys from the right, they're equal
> 	return len(l) == count
> }
>
> I can refine the comment in the code if there is question on how it works
> or whether it is correct.
>
> Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/88246?email_source=notifications&email_token=AAHADKZZN4JEAEYLLNAMUPTRFF5ELA5CNFSM4KWVOJD2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXOTYVY#discussion_r386056037>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHADK2Z7V3EBBSKQRKF47TRFF5ELANCNFSM4KWVOJDQ>
> .
>

mikedanese commented:
+1 to @BenTheElder's comment. We should avoid premature optimization that negatively impact readability and maintainability.

errordeveloper commented:
Amazing, thanks a lot @munnerz!
